File: clients/java/soroban-client/src/main/java/com/samouraiwallet/soroban/client/SorobanClient.java
package com.samouraiwallet.soroban.client;

import java.io.IOException;
import java.security.NoSuchAlgorithmException;
import java.time.LocalDateTime;
import java.util.concurrent.TimeoutException;

import com.samouraiwallet.soroban.rpc.RpcClient;
import com.samouraiwallet.soroban.User;
import com.samouraiwallet.soroban.Box;

import org.apache.commons.codec.DecoderException;

public final class SorobanClient {

    public static void main(String[] args) throws IOException {

        // parse cli
        String torProxy = "127.0.0.1";
        int torPort = 9050;
        String torUrl = "http://sorzvujomsfbibm7yo3k52f3t2bl6roliijnm7qql43bcoe2kxwhbcyd.onion";
        String role = "contributor";
        String directoryName = "samourai.soroban.private";
        int numInter = 3;

        if (args.length > 0){
            torProxy = args[0];
        }
        if (args.length > 1){
            torPort = Integer.parseInt(args[1]);
        }
        if (args.length > 2){
            torUrl = args[2];
        }
        if (args.length > 3){
            role = args[3];
        }
        if (args.length > 4){
            directoryName = args[4];
        }
        if (args.length > 5){
            numInter = Integer.parseInt(args[5]);
        }

        RpcClient rpc = new RpcClient(torProxy, torPort, torUrl);
        try {
            Boolean hashEncode = true;
            if (hashEncode) {
                directoryName = RpcClient.encodeDirectory(directoryName);   
            }

            while (true) {
                switch(role) {
                    case "initiator": SorobanClient.initiator(rpc, directoryName, hashEncode, numInter); break;
                    case "contributor": SorobanClient.contributor(rpc, directoryName, hashEncode, numInter); break;
                }
            }

        } catch (NoSuchAlgorithmException e) {
            e.printStackTrace();
        } finally {
            rpc.close();
        }
    }

    static void initiator(RpcClient rpc, String directoryName, Boolean hashEncode, int numIter) {
        User user = new User();
        try{
            System.out.println("Registering public_key");
            rpc.directoryAdd(directoryName, user.publicKey(), "long");

            String privateDirectory = String.format("%s.%s", directoryName, user.publicKey());
            if (hashEncode) {
                privateDirectory = RpcClient.encodeDirectory(privateDirectory);   
            }

            String candidatePublicKey = rpc.waitAndRemove(privateDirectory, 100);
            if (candidatePublicKey.isEmpty()){
                System.out.println("Invalid candidate_publicKey");
                
            }
            System.out.println("candidate publicKey found");
            System.out.println(candidatePublicKey);

            Box box = user.box(candidatePublicKey);
            String nextDirectory = RpcClient.encodeDirectory(user.sharedSecret(box));
            if (nextDirectory.isEmpty()){
                System.out.println("Invalid reponse next_directory");
                return;
            }
            System.out.printf("nextDirectory: %s%n", nextDirectory);

            String payload;
            System.out.println("Starting echange loop...");
            int counter = 1;
            while (numIter>0) {
                // request
                System.out.println("Sending : Ping");
                payload = box.encrypt(String.format("Ping %d %s", counter, LocalDateTime.now()));
                if (payload.isEmpty()) {
                    System.out.println("Invalid Ping message");
                    return;
                }

                rpc.directoryAdd(nextDirectory, payload, "short");
                nextDirectory = RpcClient.encodeDirectory(payload);
                if (nextDirectory.isEmpty()){
                    System.out.println("Invalid reponse next_directory");
                    return;
                }

                // response
                payload = rpc.waitAndRemove(nextDirectory, 10);
                nextDirectory = RpcClient.encodeDirectory(payload);
                if (nextDirectory.isEmpty()) {
                    System.out.println("Invalid reponse next_directory");
                    return;
                }

                String message = box.decrypt(payload);
                if (message.isEmpty()){
                    System.out.println("Invalid reponse message");
                    return;
                }
                System.out.println(String.format("Recieved: %s", message));
                counter += 1;
                numIter -= 1;
            }
                
        } catch(NoSuchAlgorithmException e) {
            e.printStackTrace();
        } catch(DecoderException e) {
            e.printStackTrace();
        } catch(TimeoutException e) {
            e.printStackTrace();
        } catch(InterruptedException e) {
            e.printStackTrace();
        } catch(IOException e) {
            e.printStackTrace();
        }
    }

    static void contributor(RpcClient rpc, String directoryName, Boolean hashEncode, int numIter) {
        User user = new User();
        try{
            System.out.println("Waiting for initiator publicKey");
            String initiatorPublicKey = rpc.waitAndRemove(directoryName, 10);
            if (initiatorPublicKey.isEmpty()){
                System.out.println("Invalid initiator publicKey");
            }
            System.out.println("Initiator publicKey found");
            String privateDirectory = String.format("%s.%s", directoryName, initiatorPublicKey);
            if (hashEncode) {
                privateDirectory = RpcClient.encodeDirectory(privateDirectory);   
            }

            System.out.println("Sending public_key");
            rpc.directoryAdd(privateDirectory, user.publicKey(), "default");

            Box box = user.box(initiatorPublicKey);
            String nextDirectory = RpcClient.encodeDirectory(user.sharedSecret(box));
            if (nextDirectory.isEmpty()){
                System.out.println("Invalid next_directory (start)");
                return;
            }

            String payload;
            System.out.println("Starting echange loop...");
            int counter = 1;
            while (numIter>0) {
                // query
                payload = rpc.waitAndRemove(nextDirectory, 10);
                if (payload.isEmpty()) {
                    System.out.println("Invalid payload (query)");
                    return;
                }
                nextDirectory = RpcClient.encodeDirectory(payload);
                if (nextDirectory.isEmpty()) {
                    System.out.println("Invalid next_directory (query)");
                    return;
                }

                String message = box.decrypt(payload);
                if (message.isEmpty()){
                    System.out.println("Invalid reponse message");
                    return;
                }

                // request
                System.out.println("Sending : Ping");
                payload = box.encrypt(String.format("Ping %d %s", counter, LocalDateTime.now()));
                if (payload.isEmpty()) {
                    System.out.println("Invalid query");
                    return;
                }
                System.out.printf("Recieved: %s%n", message);

                // response
                System.out.printf("Replying: %s%n", "Pong");
                payload = box.encrypt(String.format("Pong %d %s", counter, LocalDateTime.now()));
                if (payload.isEmpty()) {
                    System.out.println("Invalid payload (reply)");
                    return;
                }
                rpc.directoryAdd(nextDirectory, payload, "short");

                nextDirectory = RpcClient.encodeDirectory(payload);
                if (nextDirectory.isEmpty()) {
                    System.out.println("Invalid next_directory (response)");
                    return;
                }

                counter += 1;
                numIter -= 1;
            }
                
        } catch(NoSuchAlgorithmException e) {
            e.printStackTrace();
        } catch(DecoderException e) {
            e.printStackTrace();
        } catch(TimeoutException e) {
            e.printStackTrace();
        } catch(InterruptedException e) {
            e.printStackTrace();
        } catch(IOException e) {
            e.printStackTrace();
        }
    }
}

File: clients/python/soroban-client.py
import sys, time, json, datetime
import base64
import traceback

from user import User, box_encrypt, box_decrypt, box_shared
from rpc import Rpc, encode_directory, RpcTimeout, RpcCall
from cli import parse as parse_cli


def initiator(url, directory_name, hash_encode, rpc, num_iter):
    user = User()
    print("Registering public_key")
    rpc.directory_add(directory_name, user.public_key(), "long")
    
    private_directory = "%s.%s" % (directory_name, user.public_key())
    if hash_encode:
        private_directory = encode_directory(private_directory)
    if not private_directory:
        print("Invalid private_directory")
        return

    candidate_publicKey = rpc.wait_and_remove(private_directory, 100)
    if not candidate_publicKey:
        print("Invalid candidate_publicKey")
        return
    print("Contributor public_key found")

    box = user.box(candidate_publicKey)
    next_directory = encode_directory(box_shared(box))
    if not next_directory:
        return
    payload = ""

    print("Starting echange loop...")
    counter = 1
    while num_iter>0:
        # request
        print("Sending :", "Ping")
        payload = box_encrypt(box, "Ping %d %s" % (counter, datetime.datetime.now()))
        if not payload:
            print("Invalid Ping message")
            return
        rpc.directory_add(next_directory, payload, "short")
        next_directory = encode_directory(payload)
        if not next_directory:
            print("Invalid reponse next_directory")
            return

        # response
        payload = rpc.wait_and_remove(next_directory, 10)
        next_directory = encode_directory(payload)
        if not next_directory:
            print("Invalid reponse next_directory")
            return

        message = box_decrypt(box, payload)
        if not message:
            print("Invalid reponse message")
            return
        print("Recieved:", message)
        counter += 1
        num_iter -= 1


def contributor(url, directory_name, hash_encode, rpc, num_iter):
    user = User()

    initiator_publicKey = rpc.wait_and_remove(directory_name, 10)
    if not initiator_publicKey:
        print("Invalid initiator_publicKey")
        return
    print("Initiator public_key found")

    private_directory = "%s.%s" % (directory_name, initiator_publicKey)
    if hash_encode:
        private_directory = encode_directory(private_directory)

    if not private_directory:
        print("Invalid private_directory")
        return

    print("Sending public_key")
    rpc.directory_add(private_directory, user.public_key(), "default")

    box = user.box(initiator_publicKey)
    next_directory = encode_directory(box_shared(box))
    if not next_directory:
        print("Invalid next_directory (start)")
        return
    payload = ""

    print("Starting echange loop...")
    counter = 1
    while num_iter>0:
        # query
        payload = rpc.wait_and_remove(next_directory, 10)
        if not payload:
            print("Invalid payload (query)")
            return
        next_directory = encode_directory(payload)
        if not next_directory:
            print("Invalid next_directory (query)", next_directory)
            return

        message = box_decrypt(box, payload)
        if not message:
            print("Invalid query")
            return
        print("Recieved:", message)

        # response
        print("Replying:", "Pong")
        payload = box_encrypt(box, "Pong %d %s" % (counter, datetime.datetime.now()))
        if not payload:
            print("Invalid payload (reply)")
            return
        rpc.directory_add(next_directory, payload, "short")

        next_directory = encode_directory(payload)
        if not next_directory:
            print("Invalid next_directory (response)")
            return

        counter += 1
        num_iter -= 1


def main(argv):
    url, with_tor, directory_name, hash_encode, role, num_iter = parse_cli(argv)
    if hash_encode:
        directory_name = encode_directory(directory_name)
    if not directory_name:
        return

    rpc = Rpc(url, with_tor)

    while True:
        try:
            if role=="initiator":
                initiator(url, directory_name, hash_encode, rpc, num_iter)
            elif role=="contributor":
                contributor(url, directory_name, hash_encode, rpc, num_iter)
            else:
                raise Exception("Invalid role")
            print("Done.")

        except RpcTimeout as e:
            traceback.print_exc()
            print("RpcTimeout:", e)
        except RpcCall as e:
            traceback.print_exc()
            print("RpcCall:", e)
        except Exception as e:
            traceback.print_exc()
            print("Error:", e)
        except:
            traceback.print_exc()
            print("Unknown error")


if __name__ == "__main__":
   main(sys.argv[1:])
